### PR TITLE
HPCC-19946 Fix various problems optimizing LOOPs

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review.  Example issue19946 in the private regression suite shows the problems.

File: ecl/hql/hqlopt.cpp [hunk 1]
```diff
@@ -1644,6 +1644,9 @@ void CTreeOptimizer::analyseExpr(IHqlExpression * expr)
         //only look at the filename - not the parent files.
         analyseExpr(expr->queryChild(0));
         return;
+    case no_attr_expr:
```
Comment (ghalliday) [inline ecl/hql/hqlopt.cpp]: This meant that usage counts were not updated for any attributes on expressions - which caused problems for the loopfirst attribute - causing work to be duplicated.

File: ecl/hqlcpp/hqlresource.cpp [hunk 2]
```diff
@@ -602,7 +602,7 @@ class EclChildSplitPointLocator : public EclHoistLocator
             }
         case no_attr_expr:
             //Ignore internal tracking attributes e.g., _selectors_Atom
-            if (!isInternalAttributeName(expr->queryName()))
+            if (!isInternalAttributeName(expr->queryName()) || expr->queryName() == _loopFirst_Atom)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlresource.cpp]: This prevented expressions in loopfirst attributes from being executed outside the loop activity.

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 3]
```diff
@@ -9588,7 +9588,8 @@ IHqlExpression * HqlCppTranslator::getResourcedGraph(IHqlExpression * expr, IHql
     //expressions prevents child expressions from preventing fields from being removed.
     //Perform before the optimizeHqlExpression() so decisions about reducing row sizes are accurate
     traceExpression("BeforeImplicitProjectGraph", resourced);
-    resourced.setown(insertImplicitProjects(*this, resourced, false));
+    if (options.optimizeResourcedProjects)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlhtcpp.cpp]: This made debugging harder - it should be controlled by this option.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed. Looks good.

**Pull request merged**